src/value_ptr.h

```diff
@@ -61,6 +62,148 @@ bool value_ptr_equals( const value_ptr<T> &lhs, const value_ptr<T> &rhs )
     return ( !lhs && !rhs ) || ( lhs && rhs && *lhs == *rhs );
 }
 
+/**
+ * This class is essentially a copyable unique pointer, like value_ptr, except
+ * it hides the fact it is a unique_ptr. It is intended for helping make types
+ * noexcept movable by moving non-noexcept-movable types to the heap.
+ *
+ * Contractually speaking, should never be empty. However for performance reasons
+ * in eg. item, moved-from heap<> objects do not automatically reset with a new
+ * copy of the wrapped data.
+ */
+template <class T>
+struct heap : private std::unique_ptr<T> {
+    public:
+        template<typename ...Args>
+        heap( Args &&...args ) : std::unique_ptr<T>( new T{ std::forward<Args>( args )... } ) {}
+
+        // Unlike value_ptr, moves actually move and leave the moved-from heap empty.
+        // Using it again will cause a segfault without resetting it first.
+        heap( heap && ) noexcept = default;
+        heap &operator=( heap &&other ) noexcept = default;
+
+        // Like value_ptr, copies copy the rhs and leave it intact.
+        heap( heap const &rhs ) {
+            *this = rhs;
+        }
+        heap &operator=( heap const &rhs ) {
+            std::unique_ptr<T> const &rhsp = static_cast<std::unique_ptr<T> const &>( rhs );
+            // This should always be true, however, sanity says do the right thing.
+            if( rhsp ) {
+                this->reset( new T{ *rhsp } );
+            } else {
+                this->reset( new T{} );
+            }
+            return *this;
+        }
+
+        heap &operator=( T const &t ) {
+            this->reset( new T{ t } );
+            return *this;
+        }
+
+        heap &operator=( T &&t ) {
+            this->reset( new T{ std::move( t ) } );
+            return *this;
+        }
+
+        // The one weird one: since this is ultimately backed on the heap,
+        // allow operator* to get a reference to the thing.
+        // We don't allow exposing it as a pointer directly, but some places need
+        // to peel back the heap<> wrapper because otherwise template type deduction
+        // might break.
+        T &operator*() {
+            return static_cast<T &>( *this );
+        }
+
+        T const &operator*() const {
+            return static_cast<T const &>( *this );
+        }
+
+        // Implicit conversion functions
+        operator T &() {
+            return *this->get();
+        }
+
+        operator T const &() const {
```

Comment:
This is dangerous to provide because it can silently break lifetime extension in some cases, see example: https://godbolt.org/z/WTqf1h47T

Reply:
True, but no one should be stack allocating a heap<> type anyway, and you'd have to go out of your way to do the bad thing with this interface (there's no deduction guide for heap<> so there's a lot of manual typing that would need to be done to get all the conversions to happen).

I'm not sure the implicit conversion is needed anyway, I just reflectively wrote it so I'll see if I can just remove it.

Reply:
If you want to keep it you can do this instead: 
```
operator T&() & {
    return *this->get();
}
operator T&() && = delete;
operator const T&() const & {
    return *this->get();
}
operator const T&() const && = delete;```

Reply:
Nice, deleting the rvalue qualified overload. I like that.

Reply:
Alternately I can make it move the inner value when rvalue qualified.

Reply:
> operator T&() & {
>     return *this->get();
> }
> operator T&() && = delete;
> operator const T&() const & {
>     return *this->get();
> }
> operator const T&() const && = delete;

Unfortunately that has the same problems as the original, as `T&&` can bind to `const T&`.

Reply:
> Alternately I can make it move the inner value when rvalue qualified.
Unfortunately that has the same problems as the original, as `T&&` can bind to `const T&`.

Reply:
I meant make it `operator const T() &&` and return a by-value T that is move-constructed from the inner.